asmcomp/polling.ml

```diff
@@ -178,18 +183,69 @@ let instr_body handler_safe i =
   in
   instr IntSet.empty i
 
-let contains_poll instr =
-  let poll = ref false in
+let contains_any f_any instr =
+  let matches = ref [] in
   Mach.instr_iter
-      (fun i -> match i.desc with Iop (Ipoll _) -> poll := true | _ -> ())
+      (fun i -> if f_any i then matches := i :: !matches else ())
       instr;
-  !poll
+  !matches
+
+let contains_poll i =
+  let poll_match =
+    (fun i -> match i.desc with Iop (Ipoll _) -> true | _ -> false) in
+    match contains_any poll_match i with
+    | [] -> false
+    | _ -> true
+
+let contains_poll_alloc_or_calls =
+  let f_match i =
+      match i.desc with
+      | Iop(Ipoll _ | Ialloc _ | Icall_ind | Icall_imm _ |
+            Itailcall_ind | Itailcall_imm _ |
+            Iextcall { alloc = true }) -> true
+      | _ -> false
+    in
+    contains_any f_match
 
 let instrument_fundecl ~future_funcnames:_ (f : Mach.fundecl) : Mach.fundecl =
   let handler_needs_poll = polled_loops_analysis f.fun_body in
   let new_body = instr_body handler_needs_poll f.fun_body in
-  let new_contains_calls = f.fun_contains_calls || contains_poll new_body in
-  { f with fun_body = new_body; fun_contains_calls = new_contains_calls }
+  if f.fun_poll_error then begin
+    let poll_instrs = contains_poll_alloc_or_calls new_body in
+    match poll_instrs with
+    | [] -> ()
+    | _ -> raise (Error (Poll_error poll_instrs))
+  end;
+    let new_contains_calls = f.fun_contains_calls || contains_poll new_body in
+    { f with fun_body = new_body; fun_contains_calls = new_contains_calls }
```

Comment:
Indentation is strange here (and elsewhere but this is an example where it affects readability).

Reply:
Yes, you're right. Fix incoming.